### PR TITLE
flush iptable chains first and then remove them

### LINKED ISSUE #70615: kube-proxy fails to cleanup iptables chains after restart
**What happened**:

If `kube-proxy` restarts it reports

```
E1103 10:13:51.136327       1 proxier.go:532] Error removing iptables rules in ipvs proxier: error deleting chain "KUBE-MARK-MASQ": exit status 1: iptables: Too many links.
```

This is due to another chain referencing `KUBE-MARK-MASQ`. This would not happen if the chains are first all flushed and then removed.

Some examples of chains referencing `KUBE-MARK-MASQ`:

```
-A KUBE-NODEPORTS -p tcp -m comment --comment "kube-system/nginx-ingress-controller:https" -m tcp --dport 30011 -j KUBE-MARK-MASQ
-A KUBE-SEP-YJLYRFFD3HJJOLTW -s 172.20.32.131/32 -j KUBE-MARK-MASQ
-A KUBE-SERVICES ! -s 172.20.0.0/16 -d 172.31.0.10/32 -p udp -m comment --comment "kube-system/coredns:dns cluster IP" -m udp --dport 53 -j KUBE-MARK-MASQ
```

See: https://github.com/alexjx/kubernetes/blob/4ca62e4f39c59ebe66a59685a8f020315dfd3016/pkg/proxy/ipvs/proxier.go#L520

We are running in `iptables` mode. The error comes from ipvs cleanup.

**What you expected to happen**:

That all chains get cleaned up properly.

**How to reproduce it (as minimally and precisely as possible)**:

Start the proxy in `iptables` mode. Kill it. Start it again. Uncertain if specific services cause this problem. I can reproduce this in an almost empty cluster.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.0", GitCommit:"0ed33881dc4355495f623c6f22e7dd0b7632b7c0", GitTreeState:"clean", BuildDate:"2018-09-27T17:05:32Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.2", GitCommit:"17c77c7898218073f14c8d573582e8d2313dc740", GitTreeState:"clean", BuildDate:"2018-10-24T06:43:59Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration:
AWS

- OS (e.g. from /etc/os-release):
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1855.5.0
VERSION_ID=1855.5.0
BUILD_ID=2018-10-22-2305
PRETTY_NAME="Container Linux by CoreOS 1855.5.0 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
- Kernel (e.g. `uname -a`):
Linux ip-10-1-2-53.eu-central-1.compute.internal 4.14.74-coreos #1 SMP Mon Oct 22 22:12:42 UTC 2018 x86_64 Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz GenuineIntel GNU/Linux

/kind bug

### PR DESCRIPTION
**Pull request by teemow**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
/kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

while cleaning up ipvs mode. flushing iptable chains first and then
remove the chains. this avoids trying to remove chains that are still
referenced by rules in other chains.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #70615

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
NONE

```release-note

```

### CODE REVIEW

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -551,14 +551,18 @@ func cleanupIptablesLeftovers(ipt utiliptables.Interface) (encounteredError bool
 		}
 	}
 
-	// Flush and remove all of our chains.
+	// Flush and remove all of our chains. Flushing all chains before removing them also removes all links between chains first.
```
Comment (anfernee) [inline pkg/proxy/ipvs/proxier.go]: Comment should be "Flush all of our chains."

**Review by anfernee — COMMENTED** [review]
Is there a way to add some unit/integration test for this?

**Comment by thockin** [comment]
@m1093782566 This is how iptables mode works.  Any objections (other than the comment to be fixed)?

**Comment by m1093782566** [comment]
@thockin 

I think this is a good fix. I have no objections.

LGTM

**Comment by alvaroaleman** [comment]
We had some debugging fun because of this. @m1093782566 Any chance we can backport this, at least to 1.13?

**Pull request merged**